src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp

```diff
@@ -110,9 +119,18 @@ void ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::always_strong_
 template <bool CONCURRENT, bool SINGLE_THREADED>
 void ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::cld_do(CLDClosure* clds, uint worker_id) {
   if (SINGLE_THREADED) {
-    assert(SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
-    assert(Thread::current()->is_VM_thread(), "Single threaded CLDG iteration can only be done by VM thread");
-    ClassLoaderDataGraph::cld_do(clds);
+    if (CONCURRENT) {
+      if (_semaphore.try_acquire()) {
+        ShenandoahWorkerTimingsTracker timer(_phase, ShenandoahPhaseTimings::CLDGRoots, worker_id);
+        MutexLocker locker(ClassLoaderDataGraph_lock, Mutex::_no_safepoint_check_flag);
+        ClassLoaderDataGraph::cld_do(clds);
+        _semaphore.claim_all();
+      }
+    } else {
+      assert(SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
+      assert(Thread::current()->is_VM_thread(), "Can only be done by VM thread");
+      ClassLoaderDataGraph::cld_do(clds);
+    }
   } else if (_semaphore.try_acquire()) {
```

Comment:
Same as above. Also, the only difference to above is cld_do() vs always_strong_cld_do(), with *lots* of identical boilerplate. Can this be refactored into a generic method that takes the final target method as argument?

Reply:
Refactored and retested. Thanks for the suggestion.